### ISSUE TITLE
3 new coverity findings

### ISSUE THREAD

**Issue by chipitsine** [haproxy/haproxy#1416]
### Tool Name and Version

coverity

### Code Report

```plain
** CID 1464647:  Memory - illegal accesses  (USE_AFTER_FREE)
/src/listener.c: 712 in clone_listener()


________________________________________________________________________________________________________
*** CID 1464647:  Memory - illegal accesses  (USE_AFTER_FREE)
/src/listener.c: 712 in clone_listener()
706     
707      oom3:
708             free(l->name);
709      oom2:
710             free(l);
711      oom1:
>>>     CID 1464647:  Memory - illegal accesses  (USE_AFTER_FREE)
>>>     Using freed pointer "l".
712             return l;
713     }
714     
715     /* Delete a listener from its protocol's list of listeners. The listener's
716      * state is automatically updated from LI_ASSIGNED to LI_INIT. The protocol's
717      * number of listeners is updated, as well as the global number of listeners

** CID 1464646:    (NULL_RETURNS)
/src/proto_tcp.c: 651 in tcp_bind_listener()
/src/proto_tcp.c: 642 in tcp_bind_listener()
/src/proto_tcp.c: 624 in tcp_bind_listener()
/src/proto_tcp.c: 600 in tcp_bind_listener()
/src/proto_tcp.c: 663 in tcp_bind_listener()
/src/proto_tcp.c: 701 in tcp_bind_listener()
/src/proto_tcp.c: 675 in tcp_bind_listener()
/src/proto_tcp.c: 689 in tcp_bind_listener()


________________________________________________________________________________________________________
*** CID 1464646:    (NULL_RETURNS)
/src/proto_tcp.c: 651 in tcp_bind_listener()
645             }
646     #endif
647     #if defined(TCP_USER_TIMEOUT)
648             if (listener->tcp_ut) {
649                     if (setsockopt(fd, IPPROTO_TCP, TCP_USER_TIMEOUT,
650                                    &listener->tcp_ut, sizeof(listener->tcp_ut)) == -1) {
>>>     CID 1464646:    (NULL_RETURNS)
>>>     Dereferencing "msg", which is known to be "NULL".
651                             chunk_appendf(msg, "%scannot set TCP User Timeout", msg->data ? ", " : "");
652                             err |= ERR_WARN;
653                     }
654             } else
655                     setsockopt(fd, IPPROTO_TCP, TCP_USER_TIMEOUT, &zero,
656                         sizeof(zero));
/src/proto_tcp.c: 642 in tcp_bind_listener()
636                             defaultmss = sock_inet6_tcp_maxseg_default;
637     
638                     getsockopt(fd, IPPROTO_TCP, TCP_MAXSEG, &tmpmaxseg, &len);
639                     if (defaultmss > 0 &&
640                         tmpmaxseg != defaultmss &&
641                         setsockopt(fd, IPPROTO_TCP, TCP_MAXSEG, &defaultmss, sizeof(defaultmss)) == -1) {
>>>     CID 1464646:    (NULL_RETURNS)
>>>     Dereferencing "msg", which is known to be "NULL".
642                             chunk_appendf(msg, "%scannot set MSS to %d", msg->data ? ", " : "", defaultmss);
643                             err |= ERR_WARN;
644                     }
645             }
646     #endif
647     #if defined(TCP_USER_TIMEOUT)
/src/proto_tcp.c: 624 in tcp_bind_listener()
618             }
619     
620     #if defined(TCP_MAXSEG)
621             if (listener->maxseg > 0) {
622                     if (setsockopt(fd, IPPROTO_TCP, TCP_MAXSEG,
623                                    &listener->maxseg, sizeof(listener->maxseg)) == -1) {
>>>     CID 1464646:    (NULL_RETURNS)
>>>     Dereferencing "msg", which is known to be "NULL".
624                             chunk_appendf(msg, "%scannot set MSS to %d", msg->data ? ", " : "", listener->maxseg);
625                             err |= ERR_WARN;
626                     }
627             } else {
628                     /* we may want to try to restore the default MSS if the socket was inherited */
629                     int tmpmaxseg = -1;
/src/proto_tcp.c: 600 in tcp_bind_listener()
594                     *errmsg = 0;
595     
596             if (listener->state != LI_ASSIGNED)
597                     return ERR_NONE; /* already bound */
598     
599             if (!(listener->rx.flags & RX_F_BOUND)) {
>>>     CID 1464646:    (NULL_RETURNS)
>>>     Dereferencing "msg", which is known to be "NULL".
600                     chunk_appendf(msg, "%sreceiving socket not bound", msg->data ? ", " : "");
601                     goto tcp_return;
602             }
603     
604             fd = listener->rx.fd;
605     
/src/proto_tcp.c: 663 in tcp_bind_listener()
657     #endif
658     #if defined(TCP_DEFER_ACCEPT)
659             if (listener->options & LI_O_DEF_ACCEPT) {
660                     /* defer accept by up to one second */
661                     int accept_delay = 1;
662                     if (setsockopt(fd, IPPROTO_TCP, TCP_DEFER_ACCEPT, &accept_delay, sizeof(accept_delay)) == -1) {
>>>     CID 1464646:    (NULL_RETURNS)
>>>     Dereferencing "msg", which is known to be "NULL".
663                             chunk_appendf(msg, "%scannot enable DEFER_ACCEPT", msg->data ? ", " : "");
664                             err |= ERR_WARN;
665                     }
666             } else
667                     setsockopt(fd, IPPROTO_TCP, TCP_DEFER_ACCEPT, &zero,
668                         sizeof(zero));
/src/proto_tcp.c: 701 in tcp_bind_listener()
695     
696             ready = sock_accepting_conn(&listener->rx) > 0;
697     
698             if (!ready && /* only listen if not already done by external process */
699                 listen(fd, listener_backlog(listener)) == -1) {
700                     err |= ERR_RETRYABLE | ERR_ALERT;
>>>     CID 1464646:    (NULL_RETURNS)
>>>     Dereferencing "msg", which is known to be "NULL".
701                     chunk_appendf(msg, "%scannot listen to socket", msg->data ? ", " : "");
702                     goto tcp_close_return;
703             }
704     
705     #if !defined(TCP_DEFER_ACCEPT) && defined(SO_ACCEPTFILTER)
706             /* the socket needs to listen first */
/src/proto_tcp.c: 675 in tcp_bind_listener()
669     #endif
670     #if defined(TCP_FASTOPEN)
671             if (listener->options & LI_O_TCP_FO) {
672                     /* TFO needs a queue length, let's use the configured backlog */
673                     int qlen = listener_backlog(listener);
674                     if (setsockopt(fd, IPPROTO_TCP, TCP_FASTOPEN, &qlen, sizeof(qlen)) == -1) {
>>>     CID 1464646:    (NULL_RETURNS)
>>>     Dereferencing "msg", which is known to be "NULL".
675                             chunk_appendf(msg, "%scannot enable TCP_FASTOPEN", msg->data ? ", " : "");
676                             err |= ERR_WARN;
677                     }
678             } else {
679                     socklen_t len;
680                     int qlen;
/src/proto_tcp.c: 689 in tcp_bind_listener()
683                      * the kernel to create a fast open queue if there's none.
684                      */
685                     if (getsockopt(fd, IPPROTO_TCP, TCP_FASTOPEN, &qlen, &len) == 0 &&
686                         qlen != 0) {
687                             if (setsockopt(fd, IPPROTO_TCP, TCP_FASTOPEN, &zero,
688                                 sizeof(zero)) == -1) {
>>>     CID 1464646:    (NULL_RETURNS)
>>>     Dereferencing "msg", which is known to be "NULL".
689                                     chunk_appendf(msg, "%scannot disable TCP_FASTOPEN", msg->data ? ", " : "");
690                                     err |= ERR_WARN;
691                             }
692                     }
693             }
694     #endif

** CID 1464645:  Control flow issues  (UNREACHABLE)
/src/listener.c: 707 in clone_listener()


________________________________________________________________________________________________________
*** CID 1464645:  Control flow issues  (UNREACHABLE)
/src/listener.c: 707 in clone_listener()
701             HA_SPIN_INIT(&l->lock);
702             _HA_ATOMIC_INC(&jobs);
703             _HA_ATOMIC_INC(&listeners);
704             global.maxsock++;
705             return l;
706     
>>>     CID 1464645:  Control flow issues  (UNREACHABLE)
>>>     This code cannot be reached: "oom3:
free(l->name);".
707      oom3:
708             free(l->name);
709      oom2:
710             free(l);
711      oom1:
712             return l;

```
```


### Additional Information

_No response_

### Output of `haproxy -vv`

```plain
no
```

**Comment by wtarreau** [comment]
Oops! Thanks Ilya, I'll have a look at them. I really really need to stop context-switching writing one line per branch at a time and switching to a bug report, I'm seeing my code quality going down towards the average github quality when I work on issues :-(

I'll fix them before 2.5-dev10.

**Comment by wtarreau** [comment]
All fixed now, thanks a lot Ilya!
